http-netty/src/main/java/io/micronaut/http/netty/channel/NettyThreadFactory.java

```diff
@@ -38,12 +35,6 @@
  */
 @Singleton
 @Factory
-@TypeHint(value = {
-        NioServerSocketChannel.class,
-        NioSocketChannel.class
-}, typeNames = {"sun.security.ssl.SSLContextImpl$TLSContext"},
-   accessType = {TypeHint.AccessType.ALL_DECLARED_CONSTRUCTORS, TypeHint.AccessType.ALL_DECLARED_FIELDS}
-)
```

Comment:
I don't understand this change, why does the type hint for `NioServerSocketChannel` and `NioSocketChannel` no longer work?

Reply:
Because the type hint adds:
```
  "allDeclaredFields" : true,
  "allPublicMethods" : true,
```

And what we need to do is explicitly declare the constructor with `"methods":[{"name":"<init>","parameterTypes":[] }]`

I'm not 100% sure about why it fails now but the graalvm tracing agent generates the declaration of the constructor like this for both `NioServerSocketChannel` and `NioSocketChannel`. Without this change the error was something like `class not found exception, there is not public constructor for xxxx`

Reply:
Would not changing it to `TypeHint.AccessType.ALL_PUBLIC_CONSTRUCTORS` resolve the issue?

Reply:
No, that was the first thing I tried. I've tried it again just to double-check it. It fails at runtime (when using http client) with:

```
Caused by: java.lang.IllegalArgumentException: Class NioSocketChannel does not have a public non-arg constructor
	at io.netty.channel.ReflectiveChannelFactory.<init>(ReflectiveChannelFactory.java:36)
	at io.netty.bootstrap.AbstractBootstrap.channel(AbstractBootstrap.java:110)
	at io.micronaut.http.client.netty.DefaultHttpClient.<init>(DefaultHttpClient.java:287)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.buildClient(RxNettyHttpClientRegistry.java:331)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.lambda$getClient$1(RxNettyHttpClientRegistry.java:244)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.getClient(RxNettyHttpClientRegistry.java:190)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.getClient(RxNettyHttpClientRegistry.java:141)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.resolveClient(RxNettyHttpClientRegistry.java:383)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.httpClient(RxNettyHttpClientRegistry.java:186)
	at io.micronaut.http.client.netty.$RxNettyHttpClientRegistry$HttpClient0Definition.doBuild(Unknown Source)
	at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:118)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1885)
	... 103 common frames omitted
Caused by: java.lang.NoSuchMethodException: io.netty.channel.socket.nio.NioSocketChannel.<init>()
	at java.lang.Class.getConstructor0(DynamicHub.java:3082)
	at java.lang.Class.getConstructor(DynamicHub.java:1825)
	at io.netty.channel.ReflectiveChannelFactory.<init>(ReflectiveChannelFactory.java:34)
	... 115 common frames omitted
```

Reply:
Does the generated JSON actually contain `allDeclaredConstructors: true` or `allPublicConstructors: true`? Otherwise it seems like a bug in the graal config generator

Reply:
Yes, both are included:

![image](https://user-images.githubusercontent.com/559192/104019390-c45b8b00-51bb-11eb-8df4-6d5ff64a3da2.png)

Reply:
I'm asking why this no longer works, it seems a massive change if this syntax no longer works.

Reply:
Are you saying that you have asked to the GraalVM team? Maybe there has been some regression after merging something...

Reply:
Yes I am asking, this seems like it would impact many things beyond Netty if this no longer works

Reply:
I tested on a simple example and `allPublicConstructors` seems to work for the default constructor. Would it be possible to get a reproducer so I can see what happens in this case?

Reply:
@vjovanov to reproduce the issue:

- These are the changes I've done in Micronaut following Graeme's suggestions in this PR: https://github.com/micronaut-projects/micronaut-core/commit/8061cdfc7f6c18f2888c2556fbc9517b66bc91fe. Please be aware that I removed `com.sun.jndi.dns.DnsClient` from run-time initialization because the native image generation fails otherwise.
- Publish a local Micronaut 2.3.0-SNAPSHOT version with the previous changes
- `git clone https://github.com/micronaut-graal-tests/micronaut-basic-app`
- `cd micronaut-basic-app`
- `git checkout 2.3.x`
- Add `mavenLocal()` to the repositories and make sure that the micronaut 2.3.x version you published to your mavel local in used.
- Use graalvm from `release/graal-vm/21.0` branch. I've just updated and tried with the [last commit on that branch](https://github.com/oracle/graal/commit/30546ef471e6d7c50bc5cf5faea8b27d55e82974)
- `./build-native-image.sh`
- `./basic-app`

The application starts and works for some simple endpoints like `curl localhost:8080/hello/Micronaut`.

Send a request to an endpoint that uses Micronaut HTTP Client:

`curl localhost:8080/bintray/packages`

It fails with the error shows in this PR.


P.S: I've realized that this fails with 21.0.0 and not only with 21.1.0 as I though when I opened this PR, so we need to fix this before the GraalVM and Micronaut releases next week.

Reply:
@ilopmar can you setup a branch that points to the Micronaut snapshot published to JFrog OSS so that @vjovanov doesn't have to build a local snapshot. It will help a great deal.

Reply:
By default the test applications use Micronaut snapshots, but if you want to make some changes in Micronaut you need to build it locally.

Reply:
I don't think changes need to be made in Micronaut, what needs to be investigated is what has broken in GraalVM

Reply:
In that case, @vjovanov you can use the "basic-app" I showed before skipping the steps related to building Micronaut because it already uses the 2.3.0-SNAPSHOT version published to JFrog OSS.

But, as a I said before, without removing `com.sun.jndi.dns.DnsClient` from https://github.com/micronaut-projects/micronaut-core/commit/8061cdfc7f6c18f2888c2556fbc9517b66bc91fe#diff-083eb3b35618d9f42237b98d00fda177122c89d7d854f100450d478f6e473bebL17 it is not possible to build the native image. I'm not sure if that needs to be fixed in GraalVM or just removed from Micronaut. I hope @vjovanov can confirm it.

Reply:
BTW, If we really need to remove that class I can submit a PR only with that change so it gets merged and published in the next Micronaut 2.3.0-SNAPSHOT.

Reply:
I have followed the steps above and the application works for me on the tip of Graal and on the release branch for 21.0. I tried both types of configuration and it works as expected. I checked also in debugger, and all seems legit:
```
$ curl localhost:8080/bintray/packages 
{"name":"aop","linked":false}{"name":"bom","linked":false}{"name":"cli","linked":false}{"name":"core","linked":false}{"name":"discovery-client","linked":false}{"name":"function","linked":false}{"name":"function-aws","linked":false}{"name":"function-client","linked":false}{"name":"function-groovy","linked":false}{"name":"function-web","linked":false}{"name":"http","linked":false}{"name":"http-client","linked":false}{"name":"http-netty","linked":false}{"name":"http-server","linked":false}{"name":"http-server-netty","linked":false}{"name":"inject","linked":false}{"name":"inject-groovy","linked":false}{"name":"inject-java","linked":false}{"name":"jdbc","linked":false}{"name":"management","linked":false}{"name":"router","linked":false}{"name":"runtime","linked":false}{"name":"runtime-groovy","linked":false}{"name":"security","linked":false}{"name":"security-jwt","linked":false}{"name":"security-session","linked":false}{"name":"session","linked":false}{"name":"spring","linked":false}{"name":"tracing","linked":false}{"name":"validation","linked":false}{"name":"aws-common","linked":false}{"name":"cassandra","linked":false}{"name":"gorm-common","linked":false}{"name":"hibernate-gorm","linked":false}{"name":"hibernate-jpa","linked":false}{"name":"hibernate-validator","linked":false}{"name":"jdbc-dbcp","linked":false}{"name":"jdbc-hikari","linked":false}{"name":"jdbc-tomcat","linked":false}{"name":"mongo-gorm","linked":false}{"name":"mongo-reactive","linked":false}{"name":"neo4j-bolt","linked":false}{"name":"neo4j-gorm","linked":false}{"name":"netflix-archaius","linked":false}{"name":"netflix-hystrix","linked":false}{"name":"netflix-ribbon","linked":false}{"name":"rabbitmq","linked":false}{"name":"redis-lettuce","linked":false}{"name":"micrometer-core","linked":false}{"name":"micrometer-registry-atlas","linked":false}
```
Can you please try again to see if this is certainly happening?

Reply:
So, just to be clear, because it still fails for me, this is what I do:

- I've compiled again GraalVM from release branch. Commit 7894e7f7c224e0cf8f39a83c0e2f6c709b217481
- Clone the `basic-app` repo and without modifying anything I try to build the native image. This means that I'm using the latest Micronaut 2.3.0-SNAPSHOT published on JFrog OSS. The native image generation fails with:

```
[basic-app:25416]    classlist:   1,668.02 ms,  1.57 GB
[basic-app:25416]        setup:     254.43 ms,  1.57 GB
Error: Incompatible change of initialization policy for com.sun.jndi.dns.DnsClient: trying to change RUN_TIME from the command line to RERUN Contains Random references, therefore can't be included in the image heap.
Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception
Error: Image build request failed with exit status 1
```

- Now I create a new local branch from Micronaut 2.3.x branch
- I remove `com.sun.jndi.dns.DnsClient` from https://github.com/micronaut-projects/micronaut-core/blob/4e35cd2e5eff534c8d6c9d82c115a97ee366a9e1/http-netty/src/main/resources/META-INF/native-image/io.micronaut/micronaut-http-netty/native-image.properties#L18-L17
- Build Micronaut, publish it to my maven local and use it in the basic-app I cloned before. 
- I generate the native-image and this time works. I start the app and send the request `curl localhost:8080/bintray/packages`.
- The application fails with:
```
Caused by: java.lang.IllegalArgumentException: Class NioSocketChannel does not have a public non-arg constructor
	at io.netty.channel.ReflectiveChannelFactory.<init>(ReflectiveChannelFactory.java:36)
	at io.netty.bootstrap.AbstractBootstrap.channel(AbstractBootstrap.java:110)
	at io.micronaut.http.client.netty.DefaultHttpClient.<init>(DefaultHttpClient.java:287)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.buildClient(RxNettyHttpClientRegistry.java:332)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.lambda$getClient$1(RxNettyHttpClientRegistry.java:245)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.getClient(RxNettyHttpClientRegistry.java:191)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.getClient(RxNettyHttpClientRegistry.java:142)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.resolveClient(RxNettyHttpClientRegistry.java:384)
	at io.micronaut.http.client.netty.RxNettyHttpClientRegistry.httpClient(RxNettyHttpClientRegistry.java:187)
	at io.micronaut.http.client.netty.$RxNettyHttpClientRegistry$HttpClient0Definition.doBuild(Unknown Source)
	at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:118)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1885)
	... 103 common frames omitted
Caused by: java.lang.NoSuchMethodException: io.netty.channel.socket.nio.NioSocketChannel.<init>()
	at java.lang.Class.getConstructor0(DynamicHub.java:3082)
	at java.lang.Class.getConstructor(DynamicHub.java:1825)
	at io.netty.channel.ReflectiveChannelFactory.<init>(ReflectiveChannelFactory.java:34)
	... 115 common frames omitted
```


Next try:

- In `NettyThreadFactory` in Micronat I add `TypeHint.AccessType.ALL_PUBLIC_CONSTRUCTORS`, build micronaut again, use it in basic-app and check that the `reflect-config.json` file generated by Micronaut contains now `"allPublicConstructors" : true` for both `io.netty.channel.socket.nio.NioSocketChannel` and `io.netty.channel.socket.nio.NioServerSocketChannel`.
- Build the native image again, start it and send the same curl request.
- The application fails with the same error as before :-(

What am I missing?

Reply:
I had a call with Vojin and the issue is that the fat jar doesn't include the right `reflect-config.json` file generated in `micronaut-http-netty` jar file. I see the right content on IntelliJ in that jar, but when I run `./gradlew assemble` and then open the jar file, the contents of the `reflect-config.json` file are override with other configuration.

After more debug I've found that the content being overriden is coming from `micronaut-runtime` jar because somehow the directory in the final jar is wrong, even though the directory layout on the project is correct:

![image](https://user-images.githubusercontent.com/559192/104453012-14fc2b00-55a4-11eb-9342-b6b495d951f3.png)


I'm trying to figure out what's going on, but in the mean time @graemerocher do you have any clues?

Reply:
will have a look

Reply:
If I run `./gradlew runtime:clean runtime:pTML` I see the wrong directory generated.
But if I debug the GraalTypeElementVisitor while compiling the `runtime` module, the directory is correct :exploding_head: 

I've cleanup all gradle caches and it seems that it is working now. So it was probably something related gradle local caches on my side. Damn it Gradle!

I'll do more tests later today or tomorrow and if everything works I'll close this and submit a new PR removing only the class `com.sun.jndi.dns.DnsClient` from the run-time initialization.

Reply:
The snapshots of 2.3 published don't seem to have this problem with their JAR files. I can only think there is some issue with the daemon with the build you have setup?

Reply:
Yeah, I blame Gradle local caches. After removing the directory it worked again. Sorry about the noise.

I've submitted the PR with the right fix: https://github.com/micronaut-projects/micronaut-core/pull/4806